common/src/main/scala/services/teamwork/SquadService.scala

```diff
@@ -0,0 +1,3017 @@
+              case _ => ;
+            }
+
+          case _ =>
+            (pSquadOpt, action) match {
+              //the following action can be performed by the squad leader and maybe an unaffiliated player
+              case (Some(_), SelectRoleForYourself(_)) =>
+                //TODO should be possible, but doesn't work correctly due to UI squad cards not updating as expected
+//                if(squad.Leader.CharId == tplayer.CharId) {
+//                  //squad leader currently disallowed
+//                } else
+//                //the squad leader may swap to any open position; a normal member has to validate against requirements
+//                if(squad.Leader.CharId == tplayer.CharId || squad.isAvailable(position, tplayer.Certifications)) {
+//                  squad.Membership.zipWithIndex.find { case (member, _) => member.CharId == tplayer.CharId } match {
+//                    case Some((fromMember, fromIndex)) =>
+//                      SwapMemberPosition(squad.Membership(position), fromMember)
+//                      Publish(squadFeatures(squad.GUID).ToChannel, SquadResponse.AssignMember(squad, fromIndex, position))
+//                      UpdateSquadDetail(squad)
+//                    case _ => ;
+//                    //somehow, this is not our squad; do nothing, for now
+//                  }
+//                }
+//                else {
+//                  //not qualified for requested position
+//                }
+
+              //the following action can be performed by an unaffiliated player
+              case (None, SelectRoleForYourself(position)) =>
+                //not a member of any squad, but we might become a member of this one
+                GetSquad(guid) match {
+                  case Some(squad) =>
+                    if(squad.isAvailable(position, tplayer.Certifications)) {
+                      //we could join but we may need permission from the squad leader first
+                      AddInviteAndRespond(
+                        squad.Leader.CharId,
+                        RequestRole(tplayer, guid, position),
+                        invitingPlayer = 0L, //we ourselves technically are ...
+                        tplayer.Name
+                      )
+                    }
+                  case None => ;
+                  //squad does not exist? assume old local data; force update to correct discrepancy
+                }
+
+              //the following action can be performed by anyone who has tried to join a squad
+              case (_, CancelSelectRoleForYourself(_)) =>
+                val cancellingPlayer = tplayer.CharId
+                GetSquad(guid) match {
+                  case Some(squad) =>
+                    //assumption: a player who is cancelling will rarely end up with their invite queued
+                    val leaderCharId = squad.Leader.CharId
+                    //clean up any active RequestRole invite entry where we a
```

Comment:
should we log it?

Reply:
Indeed we should!  However, we need a better log standardization plan.  Many things that should be `trace` or `debug` are currently `info` and we need to rein in what does get logged.  The live server has already run into hassle with the size and number of the logs it produces.

Reply:
I wonder if we need to have more fine grained logging targets instead of the generic log severity. maybe we need to be able to select which log buffers will display to the screen. for instance, services each have their own buffer. Not something we can fix in this patch, but more thought is needed later